### ISSUE TITLE
DirichletBC.func returns a size=(0, 1) tensor causing NaN loss for the boundary condition

### ISSUE THREAD

**Issue by xuelanghanbao** [lululxvi/deepxde#1338]
Hello, I encountered an issue while using DeepXDE. When I use DirichletBC.func, it returns a tensor of size=(0, 1), which leads to a NaN value in the loss calculation for the corresponding boundary condition.

Reproducible code:

```
import numpy as np
import deepxde as dde

a=0.5
b=3.14
k=20.0

def func(x):
    return np.sin(k * x)

def gen_testdata():
    x = geomtime.random_points(2000)
    y = func(x)
    return x, y

def pde(x, y):
    dy_xx = dde.grad.hessian(y, x, i=0, j=0)
    return dy_xx + k ** 2 * y

geomtime = dde.geometry.Interval(a, b)

bc = dde.icbc.DirichletBC(geomtime, func, lambda _,on_boundary:on_boundary)
data = dde.data.PDE(geomtime, pde, bc, num_domain=300, num_test=2000, solution=func, train_distribution='uniform')
net = dde.maps.FNN([1] + [20] * 3 + [1], "tanh", "Glorot normal")
model = dde.Model(data, net)
model.compile("adam", lr=0.0001)
losshistory, train_state = model.train(iterations=40000)
```

![2](https://github.com/lululxvi/deepxde/assets/50171557/74333cbf-eaeb-4fa2-be36-9d64c1a009ca)
